core/src/main/java/com/linecorp/armeria/common/stream/TimeoutStreamMessage.java

```diff
@@ -0,0 +1,273 @@
+        private final Subscriber<? super T> delegate;
+        private final EventExecutor executor;
+        private final StreamTimeoutMode timeoutMode;
+        private final Duration timeoutDuration;
+        private final long timeoutNanos;
+        @Nullable
+        private ScheduledFuture<?> timeoutFuture;
+        @Nullable
+        private Subscription subscription;
+        private long lastEventTimeNanos;
+        private boolean completed;
+        private boolean canceled;
+
+        TimeoutSubscriber(Subscriber<? super T> delegate, EventExecutor executor, Duration timeoutDuration,
+                          StreamTimeoutMode timeoutMode) {
+            this.delegate = requireNonNull(delegate, "delegate");
+            this.executor = requireNonNull(executor, "executor");
+            this.timeoutDuration = requireNonNull(timeoutDuration, "timeoutDuration");
+            timeoutNanos = timeoutDuration.toNanos();
+            this.timeoutMode = requireNonNull(timeoutMode, "timeoutMode");
+        }
+
+        private ScheduledFuture<?> scheduleTimeout(long delay, TimeUnit unit) {
+            return executor.schedule(this, delay, unit);
+        }
+
+        void cancelSchedule() {
+            if (timeoutFuture != null && !timeoutFuture.isCancelled()) {
+                timeoutFuture.cancel(false);
+            }
+        }
+
+        @Override
+        public void run() {
+            if (timeoutMode == StreamTimeoutMode.UNTIL_NEXT) {
+                final long currentTimeNanos = System.nanoTime();
+                final long elapsedNanos = currentTimeNanos - lastEventTimeNanos;
+
+                if (elapsedNanos < timeoutNanos) {
+                    final long delayNanos = timeoutNanos - elapsedNanos;
+                    timeoutFuture = scheduleTimeout(delayNanos, TimeUnit.NANOSECONDS);
+                    return;
+                }
+            }
+            completed = true;
+            delegate.onError(new TimeoutException(
+                    String.format(TIMEOUT_MESSAGE, timeoutDuration.toMillis(), timeoutMode)));
+            subscription.cancel();
+        }
+
+        @Override
+        public void onSubscribe(Subscription s) {
+            subscription = s;
+            delegate.onSubscribe(this);
+            if (completed || canceled) {
+                return;
+            }
+            lastEventTimeNanos = System.nanoTime();
+            timeoutFuture = scheduleTimeout(timeoutNanos, TimeUnit.NANOSECONDS);
+        }
+
+        @Override
+        public void onNext(T t) {
+            if (completed || canceled) {
+                PooledObjects.close(t);
+                return;
+            }
+            switch (timeoutMode) {
+                case UNTIL_NEXT:
+                    lastEventTimeNanos = System.nanoTime();
+                    break;
+                case UNTIL_FIRST:
+                    cancelSchedule();
```

Comment:
nit: Set `timeoutFuture` to null to avoid calling `.isCancelled()` which may access a volatile field. 
```suggestion
                    cancelSchedule();
                    timeoutFuture = null;
```